lib/src/store.dart

```diff
@@ -55,14 +55,13 @@ class Store {
 
   /// Executes a given function inside a transaction
   R runInTransaction<R>(TxMode mode, R Function() fn) {
-    assert(mode == TxMode.Read, "write transactions are currently not supported"); // TODO implement
-
-    Pointer<Void> txn = bindings.obx_txn_read(_cStore);
-    checkObxPtr(txn, "failed to created transaction");
+    Pointer<Void> txn = mode == TxMode.Read ? bindings.obx_txn_read(_cStore) : bindings.obx_txn_write(_cStore);
+    checkObxPtr(txn, "failed to create transaction");
     try {
       return fn();
     } finally {
-      checkObx(bindings.obx_txn_close(txn));
+      //write transaction has to be closed with success, not sure why though
+      mode == TxMode.Write ? checkObx(bindings.obx_txn_success(txn)) : checkObx(bindings.obx_txn_close(txn));
```

Comment:
But shouldn't txn_close() also be an alternative to mark write transactions successfully and close them afterwards? As mentioned in the C bindings:  

> /// Closes (deletes) the transaction (read or write); the given OBX_txn pointer must not be used afterwards.
/// While this is the only way to release read transactions, this call is also an alternative to call obx_txn_success()
/// on write transactions.
/// In combination with obx_txn_mark_success(), this potentially commits or aborts a write transaction on the DB:
/// 1) If it's an outermost TX and all (inner) TXs were marked successful, this commits the transaction.
/// 2) If this transaction was not marked successful, this aborts the transaction (even if it's an inner TX).
/// If an error is returned (e.g. commit failed because DB is full), you can assume that the transaction was closed.
obx_err obx_txn_close(OBX_txn* txn);

Reply:
Nope, it's only for "closing" (which implies aborting a write TX if it wasn't marked successful).

Which part of the docs leads you to believe it could be used to commit by itself? Maybe the docs need to be fixed then.

Reply:
Ah I think I kinda missed the 

> In combination

 part :D

Reply:
Back to the original issue: the TX must fail if `fn()` has thrown

Reply:
But isn't that a problem of the binding itself?

Reply:
This method is the only place you actually do any operations on transactions (call obx_txn_* functions). No other part of the code has enough knowledge what's going on to be able to do handle it AFAIK. Do you have a suggestion where else it could it be implemented?

Reply:
Sorry, confused it with the recursive txn problem 🙄 I'll implement it that way :+1:

Reply:
Ok hope it works now :)

Reply:
I hope we can do better than hope :smile: What about testing it?

Reply:
Tried to come up with an other test that can force an exception, but the only thing that comes up in my mind is put/get with a non valid id... are there any other exceptions that could be forced? Or is a throw inside a `runInTransaction()` a valid prove?

Reply:
You can just `throw Exception("hello there")`

Reply:
This  
```dart
test("failing transactions", () {
    try {
      store.runInTransaction(TxMode.Write, () {
        box.putMany(simple_items);
        throw Exception("Test exception");
      });
    } on Exception {
      expect(box.count(), equals(0));
    }
  });
```
proves it then :D

Reply:
almost :tongue: :D 

it should be `finally`, instead of `on Exception`, to also run the check if the library would have eaten the exception (wouldn't throw)

Reply:
God dammit 😂 but thanks for the feedback! That really helps me :)